### PR TITLE
Support Net Core 3.0.

### PR DESCRIPTION
**Pull request by azchohfi**
Issue: #20, https://github.com/windows-toolkit/WindowsCommunityToolkit/issues/2578

## PR Type
What kind of change does this PR introduce?
- Feature

## What is the current behavior?
No support for Islands on .Net Core 3.0 WPF or WinForms projects.

## What is the new behavior?
Xaml Islands Supports .Net Core 3.0 WPF and WinForms projects.

## PR Checklist

Please check if your PR fulfills the following requirements:

- [ ] Tested code with current [supported SDKs](../readme.md#supported)
- [ ] Pull Request has been submitted to the documentation repository [instructions](..\contributing.md#docs). Link: 
- [ ] Sample in sample app has been added / updated (for bug fixes / features)
    - [ ] Icon has been created (if new sample) following the [Thumbnail Style Guide and templates](https://github.com/windows-toolkit/WindowsCommunityToolkit-design-assets)
- [ ] Tests for the changes have been added (for bug fixes / features) (if applicable)
- [X] Header has been added to all new source files (run *build/UpdateHeaders.bat*)
- [X] Contains **NO** breaking changes

### CODE REVIEW

File: Microsoft.Toolkit.Forms.UI.XamlHost/Microsoft.Toolkit.Forms.UI.XamlHost.csproj [hunk 1]
```diff
@@ -1,10 +1,9 @@
-﻿<Project Sdk="Microsoft.NET.Sdk">
+﻿<Project Sdk="Microsoft.NET.Sdk.WindowsDesktop">
 
   <PropertyGroup>
-    <TargetFramework>net462</TargetFramework>
+    <TargetFrameworks>net462;netcoreapp3.0</TargetFrameworks>
```
Comment (rido-min) [inline Microsoft.Toolkit.Forms.UI.XamlHost/Microsoft.Toolkit.Forms.UI.XamlHost.csproj]: Why net462? XamlHosts require RS5+, so I guess we can target net472. When we get into 19H1 net48 will be the default.

File: Microsoft.Toolkit.Forms.UI.XamlHost/Microsoft.Toolkit.Forms.UI.XamlHost.csproj [hunk 2]
```diff
@@ -50,6 +63,16 @@
     </Reference>
   </ItemGroup>
 
+  <ItemGroup Condition="'$(TargetFramework)' == 'netcoreapp3.0'">
+    <PackageReference Include="System.Runtime.WindowsRuntime" Version="4.6.0-preview1-27018-01" />
```
Comment (rido-min) [inline Microsoft.Toolkit.Forms.UI.XamlHost/Microsoft.Toolkit.Forms.UI.XamlHost.csproj]: We will need to review the right packages once we have the official contracts from the SDK published in NuGet

File: global.json [hunk 3]
```diff
@@ -1,5 +1,8 @@
 {
   "msbuild-sdks": {
     "MSBuild.Sdk.Extras": "1.6.55"
```
Comment (rido-min) [inline global.json]: Do we still need the extras?

File: Microsoft.Toolkit.Forms.UI.XamlHost/Microsoft.Toolkit.Forms.UI.XamlHost.csproj (hunk 1 above)
Comment (azchohfi) [reply Microsoft.Toolkit.Forms.UI.XamlHost/Microsoft.Toolkit.Forms.UI.XamlHost.csproj]: We don't want to force a dev to upgrade the app's .net framework version just because this specific feature (xaml islands) will always run on a pc that has 4.7.2 installed. We should support the min version that supports the feature. If we just update to net 472, any developer that is targeting net 462 and Win7 (for example), would be unable to use this feature, on a scenario where the dev detects if the app is running on Win7 or Win10 and gives a different UI based on that.

File: Microsoft.Toolkit.Forms.UI.XamlHost/Microsoft.Toolkit.Forms.UI.XamlHost.csproj (hunk 2 above)
Comment (azchohfi) [reply Microsoft.Toolkit.Forms.UI.XamlHost/Microsoft.Toolkit.Forms.UI.XamlHost.csproj]: Do we have an ETA on that?

File: Microsoft.Toolkit.Forms.UI.XamlHost/DpiScalingPanel.cs [hunk 4]
```diff
@@ -4,13 +4,15 @@
 
 using System;
 using System.Diagnostics;
+using foundation = Windows.Foundation;
+using uwpXaml = Windows.UI.Xaml;
```
Comment (rjmurillo) [inline Microsoft.Toolkit.Forms.UI.XamlHost/DpiScalingPanel.cs]: NIT: This naming feels strange to me

File: Microsoft.Toolkit.Forms.UI.XamlHost/Microsoft.Toolkit.Forms.UI.XamlHost.csproj [hunk 5]
```diff
@@ -50,6 +63,16 @@
...
+    <PackageReference Include="System.Runtime.WindowsRuntime.UI.Xaml" Version="4.6.0-preview1-27018-01" />
+    <Reference Include="Windows">
+      <HintPath>$(MSBuildProgramFiles32)\Windows Kits\10\UnionMetadata\10.0.17763.0\Windows.winmd</HintPath>
```
Comment (rjmurillo) [inline Microsoft.Toolkit.Forms.UI.XamlHost/Microsoft.Toolkit.Forms.UI.XamlHost.csproj]: There is a mechanism to select this in the directory.props/targets

File: Microsoft.Toolkit.Forms.UI.XamlHost/WindowsXamlHostBase.KeyboardFocus.cs [hunk 6]
```diff
@@ -85,10 +87,10 @@ protected override bool ProcessTabKey(bool forward)
                 // Temporary Focus handling for Redstone 5
```
Comment (rjmurillo) [inline Microsoft.Toolkit.Forms.UI.XamlHost/WindowsXamlHostBase.KeyboardFocus.cs]: There are 19H1 builds out in Insiders Fast--has this been fixed there?

File: Microsoft.Toolkit.Forms.UI.XamlHost/DpiScalingPanel.cs [hunk 7]
```diff
@@ -4,13 +4,15 @@
 
 using System;
 using System.Diagnostics;
+using foundation = Windows.Foundation;
```
Comment (rjmurillo) [inline Microsoft.Toolkit.Forms.UI.XamlHost/DpiScalingPanel.cs]: This has nothing to do with conversion to .NET Core 3?

File: Microsoft.Toolkit.Win32.UI.XamlHost/UWPTypeFactory.cs [hunk 8]
```diff
@@ -3,6 +3,7 @@
 // See the LICENSE file in the project root for more information.
 
 using System;
+using xaml = Windows.UI.Xaml;
```
Comment (rjmurillo) [inline Microsoft.Toolkit.Win32.UI.XamlHost/UWPTypeFactory.cs]: This is inconsistent with other using shortcuts

File: Microsoft.Toolkit.Win32.sln [hunk 9]
```diff
@@ -23,13 +23,13 @@ Project("{9A19103F-16F7-4668-BE54-9A1E7A4F7556}") = "Microsoft.Toolkit.Wpf.UI.Xa
 EndProject
 Project("{9A19103F-16F7-4668-BE54-9A1E7A4F7556}") = "Microsoft.Toolkit.Forms.UI.Controls", "Microsoft.Toolkit.Forms.UI.Controls\Microsoft.Toolkit.Forms.UI.Controls.csproj", "{9B054ABA-B5B4-42F4-8E29-97DCCF79C726}"
 EndProject
-Project("{FAE04EC0-301F-11D3-BF4B-00C04F79EFBC}") = "Microsoft.Toolkit.Sample.Forms.WebView", "Microsoft.Toolkit.Sample.Forms.WebView\Microsoft.Toolkit.Sample.Forms.WebView.csproj", "{D103E448-64B1-407C-B09E-7C61AF9F2740}"
+Project("{9A19103F-16F7-4668-BE54-9A1E7A4F7556}") = "Microsoft.Toolkit.Sample.Forms.WebView", "Microsoft.Toolkit.Sample.Forms.WebView\Microsoft.Toolkit.Sample.Forms.WebView.csproj", "{D103E448-64B1-407C-B09E-7C61AF9F2740}"
```
Comment (rjmurillo) [inline Microsoft.Toolkit.Win32.sln]: Why did the project GUIDs need to change?

File: global.json (hunk 3 above)
Comment (azchohfi) [reply global.json]: We don't! I just removed it.

File: Microsoft.Toolkit.Forms.UI.XamlHost/Microsoft.Toolkit.Forms.UI.XamlHost.csproj (hunk 1 above)
Comment (rido-min) [reply Microsoft.Toolkit.Forms.UI.XamlHost/Microsoft.Toolkit.Forms.UI.XamlHost.csproj]: I'd love to see how to use XAML adaptive coding to enable the Win7 scenario. I remember this [hack](https://github.com/qmatteoq/DesktopBridgeHelpers/blob/master/DesktopBridge.Helpers/Helpers.cs) that also required carefully coding to avoid JIT issues.

In the other hand, I think that requiring 4.7.2 is a good-enough check to avoid the app to run on older Win10 without XamlIslands.

File: Microsoft.Toolkit.Forms.UI.XamlHost/DpiScalingPanel.cs (hunk 4 above)
Comment (azchohfi) [reply Microsoft.Toolkit.Forms.UI.XamlHost/DpiScalingPanel.cs]: That's because most files do have "using System;" and this makes a conflict happen between global::System.Windows.* and global::Windows.*
Is there another way to solve this conflicts? I don't think there is. If the problem is just the naming, we could change it to something better.

File: Microsoft.Toolkit.Forms.UI.XamlHost/DpiScalingPanel.cs (hunk 7 above)
Comment (azchohfi) [reply Microsoft.Toolkit.Forms.UI.XamlHost/DpiScalingPanel.cs]: It does. The same source is being compiled with the .net framework 4.6 tools and .net core, and when using .net core with support for UWP types, we have a conflict between System.Windows.* and Windows.*

File: Microsoft.Toolkit.Win32.UI.XamlHost/UWPTypeFactory.cs (hunk 8 above)
Comment (azchohfi) [reply Microsoft.Toolkit.Win32.UI.XamlHost/UWPTypeFactory.cs]: Agreed. I'll fix it.

File: Microsoft.Toolkit.Win32.sln (hunk 9 above)
Comment (azchohfi) [reply Microsoft.Toolkit.Win32.sln]: VS updated that for me, I don't know why it did that. Should I roll back?

**Comment by azchohfi** [comment]
I'll update them this week.

**Comment by azchohfi** [comment]
Or now.

**Comment by azchohfi** [comment]
I simplified the usings so all of them now are:
```cs
using windows = Windows;
```
Then a simple case change from Windows.* to windows.* makes everything work.

**Comment by azchohfi** [comment]
The WIP is still here because I had to disable CodeAnalysis. Once this is fixed and we had a final stable version of .net core 3.0, and the nugets, then we can merge this.

File: Microsoft.Toolkit.Forms.UI.XamlHost/Microsoft.Toolkit.Forms.UI.XamlHost.csproj (hunk 5 above)
Comment (azchohfi) [reply Microsoft.Toolkit.Forms.UI.XamlHost/Microsoft.Toolkit.Forms.UI.XamlHost.csproj]: I'm using it now.

**Comment by azchohfi** [comment]
This is building with a simple tweak on MSBuild (per machine, so wont build for everybody). Should be fixed in a future version of VS2019.
The ISO for the new SDK is also still not yet available on the static link.
Aside from this 2 things, this is ready for another round of review.
@marb2000 @rjmurillo @rido-min

**Comment by azchohfi** [comment]
@marb2000 I replaced cake's MSBuild task with the DotNetCore build tasks, and now it is building on the build server, which is great news!
This is ready for review. We could improve it later with @rido-min 's PR https://github.com/windows-toolkit/Microsoft.Toolkit.Win32/pull/22, but this PR is ready to review/merge.
Once this is merged on 6.0.0, we could push a prerelease version \o/ 🎉🎉🎉

File: Directory.Build.props [hunk 10]
```diff
@@ -46,7 +46,8 @@
           ./build/Install-WindowsSDKISO.ps1 18298
         -->
         <TargetPlatformVersion>10.0.18298.0</TargetPlatformVersion>
-        <TargetPlatformMinVersion>10.0.17134.0</TargetPlatformMinVersion>
+		<!-- XAML Islands require SDK 18226 -->
+        <TargetPlatformMinVersion>10.0.18226.0</TargetPlatformMinVersion>
```
Comment (rjmurillo) [inline Directory.Build.props]: If XamlIslands requires a newer platform, then the version should be overridden in those specific `csproj` files

File: Microsoft.Toolkit.Forms.UI.Controls.WebView/WebBaseCompatibilityAdapter.cs [hunk 11]
```diff
@@ -13,8 +13,10 @@ namespace Microsoft.Toolkit.Forms.UI.Controls
     internal abstract class WebBaseCompatibilityAdapter : DependencyObject, IWebViewCompatibleAdapter
     {
         protected WebBaseCompatibilityAdapter()
-        {
-            Initialize();
+        {
+#pragma warning disable CA2214 // Do not call overridable methods in constructors
```
Comment (rjmurillo) [inline Microsoft.Toolkit.Forms.UI.Controls.WebView/WebBaseCompatibilityAdapter.cs]: Is there a bug open for this? @GraniteStateHacker received feedback to fix on initial PR but it never made it through

File: Microsoft.Toolkit.Forms.UI.Controls/Microsoft.Toolkit.Forms.UI.Controls.csproj [hunk 12]
```diff
@@ -1,17 +1,17 @@
...
     <RootNamespace>Microsoft.Toolkit.Forms.UI.Controls</RootNamespace>
     <AssemblyName>Microsoft.Toolkit.Forms.UI.Controls</AssemblyName>
-    <RunCodeAnalysis>true</RunCodeAnalysis>
 
     <Title>Windows Community Toolkit Windows Forms Controls</Title>
     <Description>This library provides XAML user controls for Windows Forms. It is part of the Windows Community Toolkit.</Description>
     <PackageTags>Win32 WindowsForms UWP XAML</PackageTags>
+  </PropertyGroup>
 
-    <!--XamlHost requires 17692 -->
```
Comment (rjmurillo) [inline Microsoft.Toolkit.Forms.UI.Controls/Microsoft.Toolkit.Forms.UI.Controls.csproj]: This is the correct place to override the `TargetPlatformMinVersion`

File: Microsoft.Toolkit.Forms.UI.XamlHost/Microsoft.Toolkit.Forms.UI.XamlHost.csproj [hunk 13]
```diff
@@ -1,20 +1,17 @@
...
     <Title>Windows Community Toolkit XAMLHost for Windows Forms</Title>
     <Description>This library provides XAML islands interop helpers for Windows Forms. It is part of the Windows Community Toolkit.</Description>
     <PackageTags>XAML Islands Win32 Forms WindowsForms WinForms XamlHost</PackageTags>
+  </PropertyGroup>
 
-    <!--
-    Needs the XamlRoot API, which is available in Insiders SDK 18298 on 19-Dec 2018
-    https://blogs.windows.com/buildingapps/2018/12/19/windows-10-sdk-preview-build-18298-available-now
-    -->
-    <TargetPlatformMinVersion>10.0.18298.0</TargetPlatformMinVersion>
```
Comment (rjmurillo) [inline Microsoft.Toolkit.Forms.UI.XamlHost/Microsoft.Toolkit.Forms.UI.XamlHost.csproj]: This is the correct place to override `TargetPlatformMinVersion` rather than DirectoryInfo.props/.targets

File: global.json [hunk 14]
```diff
@@ -1,5 +1,5 @@
 {
-  "msbuild-sdks": {
-    "MSBuild.Sdk.Extras": "1.6.55"
+  "sdk": {
+    "version": "3.0.100-preview-010046"
```
Comment (rjmurillo) [inline global.json]: There should be a script to install this (like the required SDKs)

Comment (rjmurillo) [inline global.json]: There is no 3.0.100 010046 to download from https://dotnet.microsoft.com/download/dotnet-core/3.0

**Comment by rjmurillo** [comment]
Did you run the webview unit tests run for .NET core 3? I can't try it myself since I cannot download the version of .NET 3.0 core preview used

File: global.json (hunk 14 above)
Comment (azchohfi) [reply global.json]: The build server does it with the yaml file, I'll investigate how to do it from a PS script.

File: Microsoft.Toolkit.Forms.UI.XamlHost/Microsoft.Toolkit.Forms.UI.XamlHost.csproj (hunk 13 above)
Comment (azchohfi) [reply Microsoft.Toolkit.Forms.UI.XamlHost/Microsoft.Toolkit.Forms.UI.XamlHost.csproj]: Since the Webview projs are the only ones that uses a different version, I inverted that. Webview is the only one that overrides it. Wouldn't that work?

File: Microsoft.Toolkit.Forms.UI.Controls.WebView/WebBaseCompatibilityAdapter.cs (hunk 11 above)
Comment (azchohfi) [reply Microsoft.Toolkit.Forms.UI.Controls.WebView/WebBaseCompatibilityAdapter.cs]: I don't think we have a bug opened for it, but this class is internal. Is that a real issue?

**Comment by rjmurillo** [comment]
Build script is not blocking when Windows SDK 18298 is not installed on machine

File: Directory.Build.targets [hunk 15]
```diff
@@ -1,37 +1,29 @@
...
     <When Condition="'$(IsWin32Project)' == 'true' or '$(IsWpfProject)' == 'true' or '$(IsFormsProject)' == 'true'">
       <PropertyGroup>
         <TargetPlatformVersion Condition="'$(TargetPlatformVersion)' == ''">10.0.$(DefaultTargetPlatformVersion).0</TargetPlatformVersion>
         <TargetPlatformMinVersion Condition="'$(TargetPlatformMinVersion)' == ''">10.0.$(DefaultTargetPlatformMinVersion).0</TargetPlatformMinVersion>
       </PropertyGroup>
       <ItemGroup>
+        <PackageReference Include="Microsoft.CodeAnalysis.FxCopAnalyzers" Version="2.6.2" PrivateAssets="all"/>
+      </ItemGroup>
+      <ItemGroup Condition="'$(TargetFramework)' == 'net462'">
         <Reference Include="Windows">
```
Comment (rjmurillo) [inline Directory.Build.targets]: This should be the same as L21

File: Directory.Build.targets [hunk 16]
```diff
@@ -1,37 +1,29 @@
...
           <HintPath Condition="Exists('$(MSBuildProgramFiles32)\Windows Kits\10\UnionMetadata\$(TargetPlatformMinVersion)\Windows.winmd')">$(MSBuildProgramFiles32)\Windows Kits\10\UnionMetadata\$(TargetPlatformMinVersion)\Windows.winmd</HintPath>
           <HintPath Condition="Exists('$(MSBuildProgramFiles32)\Windows Kits\10\UnionMetadata\$(TargetPlatformVersion)\Windows.winmd')">$(MSBuildProgramFiles32)\Windows Kits\10\UnionMetadata\$(TargetPlatformVersion)\Windows.winmd</HintPath>
           <Private>False</Private>
         </Reference>
       </ItemGroup>
+      <ItemGroup Condition="'$(TargetFramework)' == 'netcoreapp3.0'">
+        <PackageReference Include="System.Runtime.WindowsRuntime" Version="4.6.0-preview.18571.3" />
+        <PackageReference Include="System.Runtime.WindowsRuntime.UI.Xaml" Version="4.6.0-preview.18571.3" />
+        <Reference Include="Windows">
+          <HintPath>$(MSBuildProgramFiles32)\Windows Kits\10\UnionMetadata\$(TargetPlatformVersion)\Windows.winmd</HintPath>
```
Comment (rjmurillo) [inline Directory.Build.targets]: In my environments that have multiple SDKs installed (one for TPV and one for TPMV), resolving this reference fails. Adding the hints from L13 and L14 seem to resolve this issue.

File: Directory.Build.targets [hunk 17]
```diff
@@ -1,37 +1,29 @@
...
         <Reference Include="Windows">
           <HintPath Condition="Exists('$(MSBuildProgramFiles32)\Windows Kits\10\UnionMetadata\$(TargetPlatformMinVersion)\Windows.winmd')">$(MSBuildProgramFiles32)\Windows Kits\10\UnionMetadata\$(TargetPlatformMinVersion)\Windows.winmd</HintPath>
           <HintPath Condition="Exists('$(MSBuildProgramFiles32)\Windows Kits\10\UnionMetadata\$(TargetPlatformVersion)\Windows.winmd')">$(MSBuildProgramFiles32)\Windows Kits\10\UnionMetadata\$(TargetPlatformVersion)\Windows.winmd</HintPath>
           <Private>False</Private>
         </Reference>
       </ItemGroup>
+      <ItemGroup Condition="'$(TargetFramework)' == 'netcoreapp3.0'">
+        <PackageReference Include="System.Runtime.WindowsRuntime" Version="4.6.0-preview.18571.3" />
+        <PackageReference Include="System.Runtime.WindowsRuntime.UI.Xaml" Version="4.6.0-preview.18571.3" />
+        <Reference Include="Windows">
```
Comment (rjmurillo) [inline Directory.Build.targets]: This should be DRYd out as the inclusion of the winmd is not dependent on the target framework

File: global.json (hunk 14 above)
Comment (azchohfi) [reply global.json]: I added a ps1 for that.

**Comment by azchohfi** [comment]
Just pushed the fix of the build of the test projects. Still, not all tests are passing. I would appreciate your input @rjmurillo.

File: build/Install-DotNetCoreSDK.ps1 [hunk 18]
```diff
@@ -0,0 +1,5 @@
+﻿$globalJson = Get-Content "..\global.json" | ConvertFrom-Json
```
Comment (rjmurillo) [inline build/Install-DotNetCoreSDK.ps1]: This needs to be the relative path of the script root. For example, if you run it from the base of the repo `.\build\install-dotnetcoresdk.ps1` it fails it's looking for global.json in the parent of my current directory

**Comment by rjmurillo** [comment]
@azchohfi All tests are passing except 4, which all fail with the following:
>Message: Test method threw exception System.Resources.MissingManifestResourceException, but exception System.ArgumentException was expected. Exception message: System.Resources.MissingManifestResourceException: Could not find any resources appropriate for the specified culture or the neutral culture.  Make sure "Microsoft.Toolkit.Win32.UI.Controls.DesignerUI.resources" was correctly embedded or linked into assembly "Microsoft.Toolkit.Forms.UI.Controls.WebView" at compile time, or that all the satellite assemblies required are loadable and fully signed.

Please check that the resources are embedded properly with the new compiler.

Logs attached: [rimuri_RIMURI-W10DEV_2019-01-11_09_54_07.zip](https://github.com/windows-toolkit/Microsoft.Toolkit.Win32/files/2750637/rimuri_RIMURI-W10DEV_2019-01-11_09_54_07.zip)

File: build/Install-DotNetCoreSDK.ps1 (hunk 18 above)
Comment (azchohfi) [reply build/Install-DotNetCoreSDK.ps1]: Fixed on latest commit.

**Comment by azchohfi** [comment]
I think I found the reason... https://github.com/Microsoft/msbuild/issues/1247

File: Directory.Build.targets (hunk 17 above)
Comment (azchohfi) [reply Directory.Build.targets]: Done.

File: Directory.Build.targets (hunk 16 above)
Comment (azchohfi) [reply Directory.Build.targets]: Done. Please test again.

File: Directory.Build.targets (hunk 15 above)
Comment (azchohfi) [reply Directory.Build.targets]: Done.

**Comment by azchohfi** [comment]
I think I covered everything @rjmurillo

File: Microsoft.Toolkit.Forms.UI.Controls.WebView/Microsoft.Toolkit.Forms.UI.Controls.WebView.csproj [hunk 19]
```diff
@@ -2,7 +2,7 @@
 
   <PropertyGroup>
     <TargetFrameworks>net462;netcoreapp3.0</TargetFrameworks>
-    <RootNamespace>Microsoft.Toolkit.Forms.UI.Controls</RootNamespace>
+    <RootNamespace>Microsoft.Toolkit.Win32.UI.Controls</RootNamespace>
```
Comment (rjmurillo) [inline Microsoft.Toolkit.Forms.UI.Controls.WebView/Microsoft.Toolkit.Forms.UI.Controls.WebView.csproj]: Why is this change required?

CC @nmetulev

File: Tests/UnitTests.WebView.WinForms/FunctionalTests/Navigation/NavigateTests.cs [hunk 20]
```diff
@@ -52,18 +52,20 @@ public void NavigationFailedWithArgumentException()
         {
             WebView.Navigate(new Uri("/someresource", UriKind.Relative));
         }
-    }
-
+    }
+
+    // TestContext.TestRunResultsDirectory is only supported on .Net Framework
+#if NET462
```
Comment (rjmurillo) [inline Tests/UnitTests.WebView.WinForms/FunctionalTests/Navigation/NavigateTests.cs]: Need an alternative test for navigating to local file path for .NET Core 3.0

File: Microsoft.Toolkit.Forms.UI.Controls.WebView/Microsoft.Toolkit.Forms.UI.Controls.WebView.csproj (hunk 19 above)
Comment (azchohfi) [reply Microsoft.Toolkit.Forms.UI.Controls.WebView/Microsoft.Toolkit.Forms.UI.Controls.WebView.csproj]: It was the only way to share the resources file between the two projects, without having a different copy of them or altering the generated file. The problem is with this generated line:
https://github.com/windows-toolkit/Microsoft.Toolkit.Win32/blob/8562f121f4dc4e16fe31fb9865bc00e22216e28b/WebView.Shared/DesignerUI.Designer.cs#L42
I'm not sure this was working before, but since the compiler it self changed, this might be a regression. They had the same issue on ASP.Net Core projects (https://github.com/aspnet/AspNetCore/issues/2645), but they fixed it on ASP.Net itself, and not on dotnet core, so I opened this new issue https://github.com/dotnet/corefx/issues/34562

File: Tests/UnitTests.WebView.WinForms/FunctionalTests/Navigation/NavigateTests.cs (hunk 20 above)
Comment (azchohfi) [reply Tests/UnitTests.WebView.WinForms/FunctionalTests/Navigation/NavigateTests.cs]: That test is already [Ignored].

**Comment by rjmurillo** [comment]
@azchohfi Are you able to run `dotnet test`? There should be two test runs: one for .NET 4.6.2 and the other for .NET Core 3.0. 4.6.2 runs for me now, but Core 3.0 throws errors. Example
```
Failed   Close
Error Message:
 Initialization method Microsoft.Toolkit.Win32.UI.Controls.Test.WinForms.WebView.FunctionalTests.Close.CallCloseTwiceAfterInit.TestInitialize threw exception. System.Runtime.InteropServices.COMException: Activating a single-threaded class from MTA is not supported (Exception from HRESULT: 0x8000001D).
Stack Trace:
    at System.StubHelpers.StubHelpers.GetWinRTFactoryObject(IntPtr pCPCMD)
   at Windows.Web.UI.Interop.WebViewControlProcessOptions..ctor()
   at Microsoft.Toolkit.Win32.UI.Controls.Interop.WinRT.WebViewControlProcessOptions.ToWinRtWebViewControlProcessOptions(WebViewControlProcessOptions options) in E:\src\GitHub\windows-toolkit\Microsoft.Toolkit.Win32\WebView.Shared\Interop\WinRT\WebViewControlProcessOptions.cs:line 50
   at Microsoft.Toolkit.Win32.UI.Controls.Interop.WinRT.WebViewControlProcessOptions.ToWinRtWebViewControlProcessOptions() in E:\src\GitHub\windows-toolkit\Microsoft.Toolkit.Win32\WebView.Shared\Interop\WinRT\WebViewControlProcessOptions.cs:line 68
   at Microsoft.Toolkit.Win32.UI.Controls.Interop.WinRT.WebViewControlProcess..ctor(WebViewControlProcessOptions processOptions) in E:\src\GitHub\windows-toolkit\Microsoft.Toolkit.Win32\WebView.Shared\Interop\WinRT\WebViewControlProcess.cs:line 35
   at Microsoft.Toolkit.Forms.UI.Controls.WebView.Initialize() in E:\src\GitHub\windows-toolkit\Microsoft.Toolkit.Win32\Microsoft.Toolkit.Forms.UI.Controls.WebView\WebView.Init.cs:line 129
   at Microsoft.Toolkit.Forms.UI.Controls.WebView.System.ComponentModel.ISupportInitialize.EndInit() in E:\src\GitHub\windows-toolkit\Microsoft.Toolkit.Win32\Microsoft.Toolkit.Forms.UI.Controls.WebView\WebView.Init.cs:line 67
   at Microsoft.Toolkit.Win32.UI.Controls.Test.WinForms.WebView.FunctionalTests.Close.CallCloseTwiceAfterInit.CreateWebView() in E:\src\GitHub\windows-toolkit\Microsoft.Toolkit.Win32\Tests\UnitTests.WebView.WinForms\FunctionalTests\Close\CloseTests.cs:line 43
   at Microsoft.Toolkit.Win32.UI.Controls.Test.WinForms.WebView.FunctionalTests.WebViewContextSpecification.Given() in E:\src\GitHub\windows-toolkit\Microsoft.Toolkit.Win32\Tests\UnitTests.WebView.WinForms\FunctionalTests\WebViewContextSpecification.cs:line 63
   at Microsoft.Toolkit.Win32.UI.Controls.Test.WinForms.WebView.FunctionalTests.BlockTestStartEndContextSpecification.<Given>b__0_0() in E:\src\GitHub\windows-toolkit\Microsoft.Toolkit.Win32\Tests\UnitTests.WebView.WinForms\FunctionalTests\BlockTestStartEndContextSpecification.cs:line 24
   at Microsoft.Toolkit.Win32.UI.Controls.Test.WinForms.WebView.FunctionalTests.WebViewContextSpecification.PrintStartEnd(String className, String methodName, Action a) in E:\src\GitHub\windows-toolkit\Microsoft.Toolkit.Win32\Tests\UnitTests.WebView.WinForms\FunctionalTests\WebViewContextSpecification.cs:line 79
   at Microsoft.Toolkit.Win32.UI.Controls.Test.WinForms.WebView.FunctionalTests.BlockTestStartEndContextSpecification.Given() in E:\src\GitHub\windows-toolkit\Microsoft.Toolkit.Win32\Tests\UnitTests.WebView.WinForms\FunctionalTests\BlockTestStartEndContextSpecification.cs:line 18
   at Microsoft.Toolkit.Win32.UI.Controls.Test.WebView.Shared.ContextSpecification.TestInitialize() in E:\src\GitHub\windows-toolkit\Microsoft.Toolkit.Win32\Tests\UnitTests.WebView.Shared\ContextSpecification.cs:line 123

Standard Output Messages:


Debug Trace:

 === Starting Close.Given ===
 === Ending Close.Given ===


 === Starting Close.Cleanup ===
 WebView is not null and has not been disposed. Calling Dispose()
 === Ending Close.Cleanup ===
```

**Comment by azchohfi** [comment]
Thanks! I also investigated and opened an issue on testfx. I'll investigate your report tomorrow, but thanks for fixing it!

**Comment by azchohfi** [comment]
Ok, the issue with this is that the TestCleanup method is calling the Cleanup always on the same thread, which on .NET Core should be the STA one, and it is not. We just need the same logic we have here: https://github.com/saikrishnav/testfxSTAext/blob/master/src/STAExtensions/STAThreadManager.cs
Since that nuget package is not targetting .net core, I've copied the class and it worked locally for one test. I'm just gonna run the full test suite before commiting this, but I think we are ok now.

**Comment by azchohfi** [comment]
Ping @rjmurillo

**Comment by rjmurillo** [comment]
@azchohfi I'm getting failures in `LocalNavigationCompleted` for only one platform. Logs
[rimuri_RIMURI-W10DEV_2019-01-22_10_37_33.zip](https://github.com/windows-toolkit/Microsoft.Toolkit.Win32/files/2784164/rimuri_RIMURI-W10DEV_2019-01-22_10_37_33.zip)

**Comment by rjmurillo** [comment]
@joshholmes FYI

**Comment by azchohfi** [comment]
@rjmurillo Those tests are the ones I fixed with https://github.com/windows-toolkit/Microsoft.Toolkit.Win32/pull/7/commits/88e2a236885501c4c1e39dfa91ad4d75e5b0434c#diff-bc342634d92fe2a25e986addd75b8426L14
They executed fine for me on both frameworks.
For which platform did it failed? Is this a regression on net462 or something broke on net core 3?

**Comment by azchohfi** [comment]
```
C:\Users\alzollin\source\repos\Microsoft.Toolkit.Win32\Tests\UnitTests.WebView.WinForms>dotnet test --filter "Name=LocalNavigationCompleted"
Build started, please wait...
Build completed.

Test run for C:\Users\alzollin\source\repos\Microsoft.Toolkit.Win32\Tests\UnitTests.WebView.WinForms\bin\Debug\net462\Microsoft.Toolkit.Win32.UI.Controls.Test.WinForms.WebView.dll(.NETFramework,Version=v4.6.2)
Microsoft (R) Test Execution Command Line Tool Version 16.0.0-preview-20181205-02
Copyright (c) Microsoft Corporation.  All rights reserved.

Starting test execution, please wait...

Total tests: 3. Passed: 3. Failed: 0. Skipped: 0.
Test Run Successful.
Test execution time: 1.6336 Minutes
Test run for C:\Users\alzollin\source\repos\Microsoft.Toolkit.Win32\Tests\UnitTests.WebView.WinForms\bin\Debug\netcoreapp3.0\Microsoft.Toolkit.Win32.UI.Controls.Test.WinForms.WebView.dll(.NETCoreApp,Version=v3.0)
Microsoft (R) Test Execution Command Line Tool Version 16.0.0-preview-20181205-02
Copyright (c) Microsoft Corporation.  All rights reserved.

Starting test execution, please wait...

Total tests: 3. Passed: 3. Failed: 0. Skipped: 0.
Test Run Successful.
Test execution time: 1.6097 Minutes
```

**Comment by rjmurillo** [comment]
@azchohfi The tests only fail for me on .NET 4.6.2 and only when executing all the tests (e.g. `dotnet test -f net462`). Running them with the filter passes. The tests always pass on .NET Core 3.0.  :shipit:

**Comment by azchohfi** [comment]
Hey @rido-min, this looks like it is good to go. We'll handle the NuGet package for the WinSDK dependency in a future PR, but I think we got this one right. Could you take a look at it?

**Comment by azchohfi** [comment]
We should be good to go. Let's merge this to 6.0.0! \o/

**Comment by rido-min** [comment]
Can we hold the 6.0 release until we have the Contracts nupkg?

**Comment by azchohfi** [comment]
6.0.0 will only be merged to master and released officially only closer to 19H1. I'm confident that it will be stable by then so we could accept the PR before we ship 6.0.0.

**Comment by marb2000** [comment]
I’m Ok with Alex, we can have prereleases so app devs can test it and do prototypes as they are already demanding.
